assembly/src/parsers/io_ops.rs

```diff
@@ -90,14 +113,68 @@ pub fn parse_env(span_ops: &mut Vec<Operation>, op: &Token) -> Result<(), Assemb
 // NON-DETERMINISTIC INPUTS
 // ================================================================================================
 
-/// TODO: implement
-pub fn parse_read(_span_ops: &mut Vec<Operation>, _op: &Token) -> Result<(), AssemblyError> {
-    unimplemented!()
-}
+/// Appends the number of READ operations specified by the operation's immediate value to the span
+/// block. This pushes the specified number of items from the advice tape onto the stack.
+///
+/// "adv.push" appends the number of READ operations specified by the operation's immediate value
+/// to the span block. This pushes the specified number of items from the advice tape onto the
+/// stack.
+///
+/// "adv.loadw" appends 4 PAD operations to the span block to make space for elements
+/// from the advice tape, then appends a READW operation to read a word from the advice tape and
+/// overwrite the top 4 elements of the stack.
+///
+/// # Errors
+///
+/// Returns an AssemblyError if the instruction is invalid, malformed, missing a required
+/// parameter, or does not match the expected operation. Returns an invalid_param AssemblyError if
+/// the parameter for adv.push is not a decimal value.
+pub fn parse_adv(span_ops: &mut Vec<Operation>, op: &Token) -> Result<(), AssemblyError> {
+    // do basic validation common to all advice operations
+    validate_op_len(op, 2, 0, 1)?;
+    if op.parts()[0] != "adv" {
+        return Err(AssemblyError::unexpected_token(op, "adv.{push.n|loadw}"));
+    }
 
-/// TODO: implement
-pub fn parse_readw(_span_ops: &mut Vec<Operation>, _op: &Token) -> Result<(), AssemblyError> {
-    unimplemented!()
+    match op.parts()[1] {
+        "push" => {
+            // check that a parameter exists
+            if op.num_parts() < 3 {
+                return Err(AssemblyError::missing_param(op));
+            }
+
+            // parse the parameter value as a decimal
+            let param_idx = 2;
+            let param_value = op.parts()[param_idx];
+            // this will be the number of items to get from the advice tape
+            let n = match param_value.parse::<u64>() {
+                Ok(i) => i,
+                Err(_) => return Err(AssemblyError::invalid_param(op, param_idx)),
+            };
```

Comment:
Maybe move this out into a separate helper function? Similar to how we have `parse_element_param()` this could be `parse_int_param()`. We should also do some "bound-checking":

* Probably 0 shouldn't be a valid option (`adv.push.0` is a noop, and thus is likely a mistake).
* Maybe we should put an upper limit on this? e.g., 4 or 8?

Reply:
I added a constant and set it to 8 for now